### PR TITLE
CIRC-845: Refund lost item fees when it checked in.

### PR DESCRIPTION
**Pull request by bohdan-suprun**
https://issues.folio.org/browse/CIRC-845 - Aged to lost: Check in (effect on fees/fines) - SET COST

## Approach
* Use `loan-storage v7.1` API
* Save date and time when item was aged to lost.
* Remove delayed billing information when an aged to lost loan is checked in.
* Refund lost item fees when aged to lost item is checked in by reusing existing service.
* Assign overdue fine when aged to lost item is overdue.
* Add `writeByPath` and `getPropertyByPath` methods to allow get/write nested properties.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java [hunk 1]
```diff
@@ -73,6 +75,7 @@ public void canAgeTenItemsToLostWhenOverdueByMoreThanInterval() {
       assertThat(itemFromStorage.getJson(), isAgedToLost());
       assertThat(getLoanActions(loanFromStorage), hasAgedToLostAction());
       assertThat(loanFromStorage.getJson(), hasPatronBillingDate(loanFromStorage));
+      assertThat(loansClient.get(overdueLoan).getJson(), hasAgedToLostDate());
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java:78]: Is this the storage loan or the business logic loan? 

If it is the business logic loan, is this property defined in the relevant JSON schema?

File: src/test/java/api/loans/scenarios/CheckInAgedToLostItemTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    assertThat(loan, hasLostItemProcessingFee(isClosedCancelledItemReturned(processingFee)));
+    assertThat(loan, hasLostItemProcessingFeeActions(
+      isCancelledItemReturnedActionCreated(processingFee)));
+
+    assertThatBillingInformationRemoved(loan);
+  }
+
+  @Test
+  public void shouldAssignOverdueFine() {
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/scenarios/CheckInAgedToLostItemTest.java:75]:
```suggestion
  public void shouldChargeOverdueFine() {
```

```suggestion
  public void shouldIssueOverdueFine() {
```

```suggestion
  public void shouldBillOverdueFine() {
```

File: src/test/java/api/loans/scenarios/CheckInAgedToLostItemTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  @Before
+  public void createOwnerAndFeeTypes() {
+    feeFineOwnerFixture.cd1Owner();
+    feeFineTypeFixture.lostItemProcessingFee();
+    feeFineTypeFixture.lostItemFee();
+    feeFineTypeFixture.overdueFine();
+  }
+
+  @Test
+  public void shouldRefundPaidAmountAndCancelRemaining() {
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/scenarios/CheckInAgedToLostItemTest.java:47]:
```suggestion
  public void shouldRefundPartiallyPaidAmountAndCancelRemaining() {
```

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  DateTime getItemLostDate() {
+    if(itemStatus == DECLARED_LOST) {
+      return loan.getDeclareLostDateTime();
+    }
+
+    if(itemStatus == LOST_AND_PAID) {
+      return loan.getDeclareLostDateTime();
+    }
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java:66]: Yes, correct.

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      return loan.getAgedToLostDateTime();
+    }
+
+    return null;
+  }
+
+  boolean itemIsNotLost() {
+    return itemStatus != DECLARED_LOST && itemStatus != AGED_TO_LOST
+      && itemStatus != LOST_AND_PAID;
+  }
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java:78]: We don't have this logic in Item but in loan, did you mean loan?

File: src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java (hunk 1 above)
Comment (bohdan-suprun) [reply src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java:78]: Ah, yes. In a fact I don't need to expose the property in business logic loan, but business logic resource just uses storage representation as base and does not remove properties that are not defined in schema.

Should we create a tech debt for it maybe?

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 6]
```diff
@@ -51,10 +42,7 @@ public LostItemFeeRefundService(Clients clients) {
   public CompletableFuture<Result<CheckInContext>> refundLostItemFees(
     CheckInContext context) {
 
-    final LostItemFeeRefundContext refundFeeContext = new LostItemFeeRefundContext(
-      context.getItemStatusBeforeCheckIn(), context.getItem().getItemId(),
-      context.getLoan(), context.getLoggedInUserId(),
-      context.getCheckInServicePointId().toString());
+    final LostItemFeeRefundContext refundFeeContext = using(context);
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java]: Good point, I've renamed them to `forRenewal` and `forCheckIn`

File: src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      "    \"1\": {\n" +
+      "        \"2\": {\n" +
+      "            \"3\": {\n" +
+      "                \"4\": {\n" +
+      "                    \"5\": \"" + expectedDate + "\"\n" +
+      "                }\n" +
+      "            }\n" +
+      "        }\n" +
+      "    }\n" +
+      "}";
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java]: Yes, replaced with `writeByPath` call.

File: src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      "        \"2\": {\n" +
+      "            \"3\": {\n" +
+      "                \"4\": {}\n" +
+      "            }\n" +
+      "        }\n" +
+      "    }\n" +
+      "}";
+
+    final JsonObject object = new JsonObject(json);
+    writeByPath(object, expectedString, paths);
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java]: Yes, it creates the structure.

I've removed it.

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java]: I maybe do mean the loan (which is a little strange, as this is about the item and the item status).

Could this be consolidated into a single definition of when an item is considered lost?

File: src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java (hunk 1 above)
Comment (marcjohnson-kint) [reply src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java]: > Should we create a tech debt for it maybe?

Please do raise a technical debt issue. I believe there already is one for removing the JSON from the domain, which is part of the cause of this issue (and I regret doing it)

File: src/main/java/org/folio/circulation/domain/ItemStatus.java [hunk 9]
```diff
@@ -54,4 +54,8 @@ void setDate(String date) {
   private boolean valueMatches(String value) {
     return equalsIgnoreCase(getValue(), value);
   }
+
+  public boolean isLost() {
+    return this == DECLARED_LOST || this == AGED_TO_LOST;
+  }
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/ItemStatus.java:60]: I thought the other check (in the refund context) also included lost and paid, is this a different rule?

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+    if (itemStatus == AGED_TO_LOST) {
+      return loan.getAgedToLostDateTime();
+    }
+
+    return null;
+  }
+
+  boolean shouldRefundFeesForItem() {
+    return itemStatus.isLost() || itemStatus == LOST_AND_PAID;
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java:76]: Does this mean that an item is not considered lost if it is `Lost and paid`?

File: src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import static org.folio.circulation.support.JsonPropertyWriter.writeByPath;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.allOf;
+
+import org.junit.Test;
+
+import io.vertx.core.json.JsonObject;
+
+public class JsonPropertyWriterTest {
+
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java:13]: Personally, I would remove this blank line

File: src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  @Test
+  public void shouldCreateMissingObjects() {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java:15]:
```suggestion
  public void shouldCreateMissingObjectsInThePath() {
```

File: src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  @Test
+  public void shouldWriteStringByPath() {
+    final String[] paths = {"1", "2", "3", "4", "5"};
+    final String expectedString = "A string";
+
+    final JsonObject object = new JsonObject();
+    writeByPath(object, expectedString, paths);
+
+    assertThat(object, hasJsonPath("1.2.3.4.5", "A string"));
+  }
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java:43]: As this example is a simpler example than the one above, I might be tempted to put it first in the class

File: src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  public void shouldReturnDateTimePropertyByPath() {
+    final String[] paths = {"1", "2", "3", "4", "5"};
+    final DateTime expectedDate = DateTime.parse("2020-12-12T20:00:00.123Z");
+
+    final JsonObject object = new JsonObject();
+    writeByPath(object, expectedDate, paths);
+
+    final DateTime actualValue = getDateTimePropertyByPath(object, paths);
+    assertThat(actualValue, is(expectedDate));
+  }
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java:37]: Similarly, I might be tempted to move this to the beginning of the tests as it is the simpler case

File: src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.joda.time.DateTime;
+import org.junit.Test;
+
+import io.vertx.core.json.JsonObject;
+
+public class JsonPropertyFetcherTest {
+
+  @Test
+  public void shouldReturnNullIfSomeObjectsNotExistInPath() {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java:17]: Is there another scenario where all of the objects in the path exist however the property does not?

Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java:17]:
```suggestion
  public void shouldReturnNullWhenObjectsInThePathAreNotPresent() {
```

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java]: In most cases by lost we mean either `Aged to lost` or `Declared lost` status, for example for renewal, due date change, check out, etc.

`Lost and paid` status is also can be defined as lost, but in this case the "lost" situation has already been resolved.

Maybe we can define `lost and not resolved = Aged to lost or Declared lost` and `lost and resolved = Lost and paid`.

In most places we need to check if item is `lost and not resolved` but for fees refund we need to check if item is `lost and not resolved` OR `lost and resolved`.

File: src/main/java/org/folio/circulation/domain/ItemStatus.java (hunk 9 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/ItemStatus.java:60]: See this comment https://github.com/folio-org/mod-circulation/pull/655#discussion_r483613499

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java (hunk 10 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java:76]: See https://github.com/folio-org/mod-circulation/pull/655#discussion_r483613499

File: src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java (hunk 13 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/support/JsonPropertyWriterTest.java:43]: I just realized that it is a duplicate of the 1st test. I'll remove it.

File: src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java (hunk 1 above)
Comment (bohdan-suprun) [reply src/test/java/api/loans/agetolost/ScheduledAgeToLostApiTest.java]: Here it is https://issues.folio.org/browse/CIRC-915

File: src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java (hunk 15 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/support/JsonPropertyFetcherTest.java]: I've added one, thanks.

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundContext.java]: By all means rename the current method and add a new method to create that distinction

**Review by marcjohnson-kint — APPROVED** [review]

**Pull request merged**